// @microsoft/fast-components/dist/esm/text-area/index.js
import { __decorate } from "tslib";
import { attr } from "@microsoft/fast-element";
import { TextArea as FoundationTextArea, textAreaTemplate as template, } from "@microsoft/fast-foundation";
import { textAreaStyles as styles } from "./text-area.styles";
/**
 * @internal
 */
export class TextArea extends FoundationTextArea {
    /**
     * @internal
     */
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = "outline";
        }
    }
}
__decorate([
    attr
], TextArea.prototype, "appearance", void 0);
/**
 * A function that returns a {@link @microsoft/fast-foundation#TextArea} registration for configuring the component with a DesignSystem.
 * Implements {@link @microsoft/fast-foundation#textAreaTemplate}
 *
 *
 * @public
 * @remarks
 * Generates HTML Element: \<fast-text-area\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
export const fastTextArea = TextArea.compose({
    baseName: "text-area",
    baseClass: FoundationTextArea,
    template,
    styles,
    shadowOptions: {
        delegatesFocus: true,
    },
});
/**
 * Styles for TextArea
 * @public
 */
export const textAreaStyles = styles;
