// @microsoft/fast-components/dist/esm/text-field/index.js
import { __decorate } from "tslib";
import { attr } from "@microsoft/fast-element";
import { TextField as FoundationTextField, textFieldTemplate as template, } from "@microsoft/fast-foundation";
import { textFieldStyles as styles } from "./text-field.styles";
/**
 * @internal
 */
export class TextField extends FoundationTextField {
    /**
     * @internal
     */
    connectedCallback() {
        super.connectedCallback();
        if (!this.appearance) {
            this.appearance = "outline";
        }
    }
}
__decorate([
    attr
], TextField.prototype, "appearance", void 0);
/**
 * A function that returns a {@link @microsoft/fast-foundation#TextField} registration for configuring the component with a DesignSystem.
 * Implements {@link @microsoft/fast-foundation#textFieldTemplate}
 *
 *
 * @public
 * @remarks
 * Generates HTML Element: \<fast-text-field\>
 *
 * {@link https://developer.mozilla.org/en-US/docs/Web/API/ShadowRoot/delegatesFocus | delegatesFocus}
 */
export const fastTextField = TextField.compose({
    baseName: "text-field",
    baseClass: FoundationTextField,
    template,
    styles,
    shadowOptions: {
        delegatesFocus: true,
    },
});
/**
 * Styles for TextField
 * @public
 */
export const textFieldStyles = styles;
